---

# Task 1: Familiarize Yourself with the Dataset

There are two main subtasks:  
1. Load the four data files that are part of this dataset.  
2. Explore each dataset and define the data types of each column. Describe the contents of each column, including the nature of the data values. Follow a similar approach as demonstrated in the lecture and tutorial. Discuss the shape of the data, value ranges, and possible categories. Convert any data into a more appropriate format if needed.  

Add as many code and markdown cells as necessary below. Typically, you should perform one type of investigation per code cell with output, and then comment on the results in a markdown cell below it.  

After completing the analysis, write a concise summary in the **Task 1: summary report** cell at the end. The final report should succinctly summarize the main characteristics of each data file, how they relate to each other, and any notable features.


In [65]:
# this is your first code cell, add more cells for code and mardown as needed
# imports
import pandas as pd
from IPython.display import display
from datetime import datetime

In [66]:
# TASK 1a: SESSIONS.CSV
df_sessions = pd.read_csv("data/sessions.csv")
display(df_sessions)
display (f"The shape of the dataframe is: {df_sessions.shape}")
columns = [*df_sessions.columns] # this stringifies the items, sep in the list, separating them with a comma (https://www.geeksforgeeks.org/python/print-lists-in-python-4-different-ways/)
print(f"Columns are: {columns}")
print("\nGiven below are descriptive stats on the numerical columns\n")
df_sessions.describe()

,session_id,viewer_id,streamer_id,stream_id,started_at,ended_at,duration_mins,chat_messages_sent,bits_cheered,followed_during,subscribed_during
0,SES000001,VWR00938,STR0055,SM00860,2024-01-18 19:16:00,2024-01-18 20:28:00,72,0,0,False,False
1,SES000002,VWR00804,STR0034,SM00542,2024-01-04 17:11:00,2024-01-04 19:41:00,150,11,0,False,False
2,SES000003,VWR00704,STR0008,SM00124,2024-02-15 15:11:00,2024-02-15 16:53:00,102,5,0,False,False
3,SES000004,VWR00926,STR0024,SM00373,2024-01-18 19:59:00,2024-01-18 20:56:00,57,3,9663,False,False
4,SES000005,VWR00883,STR0008,SM00120,2024-03-19 20:08:00,2024-03-19 20:22:00,14,7,0,False,False
...,...,...,...,...,...,...,...,...,...,...,...
22820,SES022821,VWR00856,STR0005,SM00083,2024-01-16 18:51:00,2024-01-16 22:17:00,206,31,0,False,False
22821,SES022822,VWR00985,STR0013,SM00200,2024-01-08 20:28:00,2024-01-08 20:52:00,24,1,0,False,False
22822,SES022823,VWR00645,STR0045,SM00711,2024-03-26 04:15:00,2024-03-26 04:58:00,43,1,0,False,False
22823,SES022824,VWR00291,STR0005,SM00078,2024-01-11 20:02:00,2024-01-11 21:01:00,59,2,0,False,False


'The shape of the dataframe is: (22825, 11)'

Columns are: ['session_id', 'viewer_id', 'streamer_id', 'stream_id', 'started_at', 'ended_at', 'duration_mins', 'chat_messages_sent', 'bits_cheered', 'followed_during', 'subscribed_during']

Given below are descriptive stats on the numerical columns



,duration_mins,chat_messages_sent,bits_cheered
count,22825.000000,22825.000000,22825.000000
mean,59.051303,6.442629,398.425323
std,60.383979,46.576309,1571.528658
min,0.000000,0.000000,0.000000
25%,15.000000,0.000000,0.000000
50%,39.000000,1.000000,0.000000
75%,83.000000,5.000000,0.000000
max,443.000000,1987.000000,10000.000000


<details>
<summary> Task 1a (sessions.csv): click to expand </summary>

### Task 1a: Summarizing sessions.csv

Since every file is converted to a dataframe and is being treated like a table, I will often use the term table and file interchangably (within reason) in this assignment.

#### Overview
This table contains information about individual viewing sesssions.

#### Table shape
- **Rows**: There are 22825 rows (22824 entries, since one row is for the column headings.)
- **Columns**: There are 11 columns. They are described below.

##### Columns description
|Column Header | Decription | Type of data | Comments |
|--------------|------------|--------------|----------|
| `session_id` | unique identifier for the inidividual session | Categorical (nominal) | N/A |
| `viewer_id` | unique identifer for the viewer (consistent across other files) | Cateogrical (nominal) | N/A |
| `streamer_id` | unique identifer for the viewer (consistent across other files) | Categorical (nominal) | N/A |
| `stream_id` | unique identifer for the stream session (consistent across other files) | Categorical (nominal) | N/A |
| `started_at` | Time and date the stream started. Format: YYYY-MM-DD HH:MM:SS | Numerical (interval) | I will create another column with this in unix timestamp to make calculations easier |
| `ended_at` | Time and date the stream ended. Format: YYYY-MM-DD HH:MM:SS | Numerical (interval) | I will create another column with this in unix timestamp to make calculations easier
| `duration_mins` | The length of the stream in minutes. | Numerical (ratio) | N/A |
| `chat_messages_sent` | The number of messages set on the chat | Numerical (ratio) | N/A |
| `bits_cheered` | The number bit cheers sent (a twitch currency) | Numerical (ratio) | N/A |
| `followed_during` | Boolean representing if the viewer followed the streamer during the session | Categorical (boolean) | N/A |
| `subscribed_during` | Boolean representing if the viewer subscribed the streamer during the session | Categorical (boolean) | N/A |

</details>

Below, I will convert the `started_at` and `ended_at` time to unix timestamp

In [ ]:
# Task 1a: Continued, creating new columns with unix_timestamp
df_sessions["started_at_unix"] = (pd.to_datetime(df_sessions["started_at"]) - pd.Timestamp("1970-01-01")) // pd.Timedelta("1s")
started_at_unix = df_sessions.pop("started_at_unix")
df_sessions.insert(loc=5, column="started_at_unix_time", value=started_at_unix)

df_sessions["ended_at_unix"] = (pd.to_datetime(df_sessions["ended_at"]) - pd.Timestamp("1970-01-01")) // pd.Timedelta("1s")
ended_at_unix = df_sessions.pop("ended_at_unix")
df_sessions.insert(loc=7, column="ended_at_unix_time", value=ended_at_unix)
display(df_sessions)

,session_id,viewer_id,streamer_id,stream_id,started_at,started_at_unix_time,ended_at,ended_at_unix_time,duration_mins,chat_messages_sent,bits_cheered,followed_during,subscribed_during
0,SES000001,VWR00938,STR0055,SM00860,2024-01-18 19:16:00,1705605360,2024-01-18 20:28:00,1705609680,72,0,0,False,False
1,SES000002,VWR00804,STR0034,SM00542,2024-01-04 17:11:00,1704388260,2024-01-04 19:41:00,1704397260,150,11,0,False,False
2,SES000003,VWR00704,STR0008,SM00124,2024-02-15 15:11:00,1708009860,2024-02-15 16:53:00,1708015980,102,5,0,False,False
3,SES000004,VWR00926,STR0024,SM00373,2024-01-18 19:59:00,1705607940,2024-01-18 20:56:00,1705611360,57,3,9663,False,False
4,SES000005,VWR00883,STR0008,SM00120,2024-03-19 20:08:00,1710878880,2024-03-19 20:22:00,1710879720,14,7,0,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...
22820,SES022821,VWR00856,STR0005,SM00083,2024-01-16 18:51:00,1705431060,2024-01-16 22:17:00,1705443420,206,31,0,False,False
22821,SES022822,VWR00985,STR0013,SM00200,2024-01-08 20:28:00,1704745680,2024-01-08 20:52:00,1704747120,24,1,0,False,False
22822,SES022823,VWR00645,STR0045,SM00711,2024-03-26 04:15:00,1711426500,2024-03-26 04:58:00,1711429080,43,1,0,False,False
22823,SES022824,VWR00291,STR0005,SM00078,2024-01-11 20:02:00,1705003320,2024-01-11 21:01:00,1705006860,59,2,0,False,False


In [ ]:
# TASK 1b: STREAMERS.CSV
df_streamers = pd.read_csv("data/streamers.csv")
display(df_streamers)
display (f"The shape of the dataframe is: {df_streamers.shape}")
columns_streamers = [*df_streamers.columns] # this stringifies the items, sep in the list, separating them with a comma (https://www.geeksforgeeks.org/python/print-lists-in-python-4-different-ways/)
print(f"Columns are: {columns_streamers}")

print("\nGiven below are descriptive stats on the numerical columns\n")
df_streamers.describe()

,streamer_id,streamer_name,category,language,partner_status,total_followers,avg_concurrent_viewers
0,STR0001,xQc,Just Chatting,English,True,11500000,138706
1,STR0002,Pokimane,Just Chatting,English,True,-25551,71660
2,STR0003,HasanAbi,Just Chatting,English,True,2600000,34220
3,STR0004,Amouranth,Just Chatting,English,True,6200000,68607
4,STR0005,Mizkif,Just Chatting,English,True,2100000,6771
...,...,...,...,...,...,...,...
71,STR0072,BobRoss_Twitch,Creative,English,True,940000,6872
72,STR0073,Mizkif_chess,Creative,English,False,450000,5892
73,STR0074,Pituquistream,Just Chatting,Spanish,True,780000,3932
74,STR0075,ProxyPipe,MOBA,English,False,230000,634


'The shape of the dataframe is: (76, 7)'

Columns are: ['streamer_id', 'streamer_name', 'category', 'language', 'partner_status', 'total_followers', 'avg_concurrent_viewers']


<details>
<summary> Task 1b (streamers.csv): click to expand </summary>

### Task 1b: Summarizing streamers.csv

Since every file is converted to a dataframe and is being treated like a table, I will often use the term table and file interchangably (within reason) in this assignment.

#### Overview
This table contains information about each streamer.

#### Table shape
- **Rows**: There are 76 rows (75 entries, since one row is for the column headings). Thus it means we have info on 75 streamers. 
- **Columns**: There are 7 columns. They are described below.

##### Columns description
|Column Header | Decription | Type of data | Comments |
|--------------|------------|--------------|----------|
| `streamer_id` | Unique identifier for the streamer. | Categorical (nominal) | N/A |
| `streamer_name` | The name/display name of the streamer. | Cateogrical (nominal) | N/A |
| `category` | Type of stream/category/genre. | Categorical (nominal) | N/A |
| `language` | The language they stream in. | Categorical (nominal) | There are some missing values | 
| `partner_status` | Whether they are a twitch partner. | Categorical (boolean) | N/A |
| `total_followers` | Total number of followers of this streamer. | Numerical (ratio) | Contains some negative values. Those are problematic datapoints |
| `avg_concurrent_viewers` | Average number of viewers that watch streams of this streamer at a time. | Numerical (ratio) | N/A |

</details>

In [ ]:
# TASK 1c: STREAMS.CSV
df_streams = pd.read_csv("data/streams.csv")
display(df_streams)
display (f"The shape of the dataframe is: {df_streams.shape}")
columns_streams = [*df_streams.columns] # this stringifies the items, sep in the list, separating them with a comma (https://www.geeksforgeeks.org/python/print-lists-in-python-4-different-ways/)
print(f"Columns are: {columns_streams}")

print("\nGiven below are descriptive stats on the numerical columns\n")
df_streams.describe()

,stream_id,streamer_id,started_at,ended_at,stream_duration_hrs,category,peak_viewers,title_has_hype_word,was_raid
0,SM00001,STR0001,2024-02-14 18:00:00,2024-02-14 21:00:00,3.0,Just Chatting,108989.0,False,True
1,SM00002,STR0001,2024-01-28 18:00:00,2024-01-28 21:00:00,3.0,Just Chatting,51635.0,True,False
2,SM00003,STR0001,2024-01-31 09:00:00,2024-01-31 16:00:00,7.0,Just Chatting,212127.0,False,False
3,SM00004,STR0001,2024-03-03 14:00:00,2024-03-03 21:00:00,7.0,Just Chatting,186745.0,False,True
4,SM00005,STR0001,2024-01-24 18:00:00,2024-01-24 21:00:00,3.0,Just Chatting,87723.0,False,True
...,...,...,...,...,...,...,...,...,...
1194,SM01195,STR0076,2024-01-31 19:00:00,2024-01-31 23:00:00,4.0,FPS,1421.0,False,False
1195,SM01196,STR0076,2024-03-20 19:00:00,2024-03-20 21:30:00,2.5,FPS,2032.0,True,False
1196,SM01197,STR0076,2024-01-01 14:00:00,2024-01-01 21:00:00,7.0,FPS,1967.0,False,False
1197,SM01198,STR0076,2024-01-02 20:00:00,2024-01-02 23:00:00,3.0,FPS,2041.0,False,False


'The shape of the dataframe is: (1199, 9)'

Columns are: ['stream_id', 'streamer_id', 'started_at', 'ended_at', 'stream_duration_hrs', 'category', 'peak_viewers', 'title_has_hype_word', 'was_raid']


<details>
<summary> Task 1c (streams.csv): click to expand </summary>

### Task 1c: Summarizing streams.csv

Since every file is converted to a dataframe and is being treated like a table, I will often use the term table and file interchangably (within reason) in this assignment.

#### Overview
This table contains information about each stream.

#### Table shape
- **Rows**: There are 1199 rows (1198 entries, since one row is for the column headings). Thus it means we have info on 1198 streams. 
- **Columns**: There are 9 columns. They are described below.

##### Columns description
|Column Header | Decription | Type of data | Comments |
|--------------|------------|--------------|----------|
| `stream_id` | Unique identifier for the streamer. | Categorical (nominal) | N/A |
| `streamer_id` | The name/display name of the streamer. | Cateogrical (nominal) | N/A |
| `started_at` | Type of stream/category/genre. | Categorical (nominal) | Will convert to unix timestamp for calculation |
| `ended_at` | The language they stream in. | Categorical (nominal) | Will convert to unix timestamp for calculation | 
| `stream_duration_hrs` | Whether they are a twitch partner. | Categorical (boolean) | N/A |
| `category` | Total number of followers of this streamer. | Numerical (ratio) | Contains some negative values. Those are problematic datapoints |
| `peak_viewers` | Average number of viewers that watch streams of this streamer at a time. | Numerical (ratio) | N/A |
| `title_has_hype_word` | Whether the title has a hype word | Categorical (boolean) | N/A |
| `was_raid` | Was it a raid from another stream? | Categorical (boolean) | N/A |

</details>

Below, I wil convert the times to unix timestamp

In [70]:
# Task 1c: Continued, creating new columns with unix_timestamp
df_streams["started_at_unix"] = (pd.to_datetime(df_streams["started_at"]) - pd.Timestamp("1970-01-01")) // pd.Timedelta("1s")
started_at_unix_streams = df_streams.pop("started_at_unix")
df_streams.insert(loc=3, column="started_at_unix_time", value=started_at_unix_streams)

df_streams["ended_at_unix"] = (pd.to_datetime(df_streams["ended_at"]) - pd.Timestamp("1970-01-01")) // pd.Timedelta("1s")
ended_at_unix_streams = df_streams.pop("ended_at_unix")
df_streams.insert(loc=5, column="ended_at_unix_time", value=ended_at_unix)
display(df_streams)

,stream_id,streamer_id,started_at,started_at_unix_time,ended_at,ended_at_unix_time,stream_duration_hrs,category,peak_viewers,title_has_hype_word,was_raid
0,SM00001,STR0001,2024-02-14 18:00:00,1707933600,2024-02-14 21:00:00,1705609680,3.0,Just Chatting,108989.0,False,True
1,SM00002,STR0001,2024-01-28 18:00:00,1706464800,2024-01-28 21:00:00,1704397260,3.0,Just Chatting,51635.0,True,False
2,SM00003,STR0001,2024-01-31 09:00:00,1706691600,2024-01-31 16:00:00,1708015980,7.0,Just Chatting,212127.0,False,False
3,SM00004,STR0001,2024-03-03 14:00:00,1709474400,2024-03-03 21:00:00,1705611360,7.0,Just Chatting,186745.0,False,True
4,SM00005,STR0001,2024-01-24 18:00:00,1706119200,2024-01-24 21:00:00,1710879720,3.0,Just Chatting,87723.0,False,True
...,...,...,...,...,...,...,...,...,...,...,...
1194,SM01195,STR0076,2024-01-31 19:00:00,1706727600,2024-01-31 23:00:00,1708543680,4.0,FPS,1421.0,False,False
1195,SM01196,STR0076,2024-03-20 19:00:00,1710961200,2024-03-20 21:30:00,1710093060,2.5,FPS,2032.0,True,False
1196,SM01197,STR0076,2024-01-01 14:00:00,1704117600,2024-01-01 21:00:00,1707751680,7.0,FPS,1967.0,False,False
1197,SM01198,STR0076,2024-01-02 20:00:00,1704225600,2024-01-02 23:00:00,1710960360,3.0,FPS,2041.0,False,False


### Task 1 Summary Report

Write your report here, summarizing the work done in the cells above.

# Task 2: Data Cleaning, Diagnostics, and Corrections

Identify and resolve issues in the dataset. These can be of different kinds, such as:
- Look for duplicates or implausible values (e.g., users with negative followers, invalid age groups). Thoroughly document your exploration of the data. Decide whether these are outliers or the result of a systemic problem. If the latter, resolve the problem here; if they are outliers, we will address those later.  
- Define at least three heuristics (common-sense rules) to flag unusual entries (e.g., someone watching dozens of streams per day). Discuss whether what you found are errors or artifacts.  

I will repeat the instructions here, but this is expected for each task: each code cell should focus on a specific analysis. You should run the code, and in the markdown below each code cell and its result, provide a detailed interpretation of what you have uncovered in that step. After completing the analyses, summarize the key findings in the indicated summary report cell below.


In [71]:
# start code for Task 2 here

### Task 2 summary report

# Task 3: Descriptive Statistics and Visualizations

- Create summary statistics and visualizations to understand user characteristics  
- Create summary statistics for streams data overall, and summaries by age group, account tire, or peak viewers  
- Visualize the distribution of streams durations, popular categories,   
- Show category-specific information and compare streamers based on categories  

In [72]:
#start code for Task 3 here

### Task 3 Summary report
summarize your findings here. 

# Task 4: Dealing with Outliers and Missing Values

- Are there any missing values or outliers in the original data? Perform the analysis and report your findings.  
- Are there any outliers or missing values in the derived data? Perform the analysis and report your findings.  
- If there are no outliers or missing values, introduce 3–5% of missing values in a selected column and 1–2% of outliers in another column.  
- Experiment with two strategies for imputing missing values and two strategies for handling outliers. Show how each strategy affects the characteristics of the data.


In [73]:
#start code for Task 4 here

### Task 4 Summary report

# Task 5: Correlation and Behavioral Patterns

- Compute pairwise correlations between user features and the number of check-ins  
- Visualize the results using a correlation matrix (use a coolwarm color scheme and proper scaling)  
- Try Spearman rank correlation and explain when it is more appropriate than Pearson

In [74]:
# tast 5 code starts here

### Task 5 Summary report

# Task 6: Compare Behavior Across User Segments

- Segment users into 2–3 groups (e.g., by age bracket, account tire)  
- Compare their viewing behavior over time using boxplots or dot plots  
- Identify any outliers in group behavior and investigate them

In [75]:
# task 6 code start here

### Task 6 Summary report

# Task 7: Your Own EDA Questions

Formulate and explore two hypotheses of your own:  
- Propose questions based on curiosity (e.g., “Does a user's variety of categories influence morning vs. evening or session duration?”)  
- Use summary statistics and visualizations to answer each question  
- Briefly discuss what additional data would help you answer the questions more effectively


In [76]:
# task 7 code start here

### Task 7 Summary report

# Task 8: EDA Presentation (Insight Report)

Summarize your findings in a polished format:  
- Create 3–4 slides or a single-page report highlighting your most important insights  
- Include at least 2 charts and interpret them using clear, plain language  
- Apply Tufte’s principles: minimize chartjunk, maximize data-to-ink ratio, and ensure clear labeling  